[SDK] Add AdaptingCircularBufferCounter for exponential histograms

euroelessar opened the pull request:
Updates #1391 

## Changes

Add AdaptingCircularBufferCounter, which will be used for buckets in exponential histograms.
This PR is mostly a rewrite of corresponding opentelemetry-java implementation and tests.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [x] Unit tests have been added
* [ ] Changes in public API reviewed

lalitb commented on sdk/src/metrics/data/circular_buffer.cc:
```diff
@@ -0,0 +1,182 @@
+  template <typename T>
+  void operator()(std::vector<T> &backing)
+  {
+    std::fill(backing.begin(), backing.end(), 0);
```
Not sure, but you may want to typecast to `static_cast<T>(0)` to get rid of CI failure ?

lalitb reviewed (approved):
Thanks for the contribution. Nicely done.

esigo reviewed (approved):
LGTM
thanks for the PR :)

marcalff reviewed (approved):
LGTM.

About the CI failure:

* check CI build `CI / CMake msvc (maintainer mode) (pull_request)`
* check the logs, look for `circular_buffer.cc`, look for `error C2220`
* see suggested fix from Lalit

Thanks for the PR.

euroelessar commented:
Thanks for the review.
I've been looking at coming usage of these containers in histograms themselves and had to make couple of adjustments:
1. changed index to be signed (as allowed by the spec)
2. added moving constructors/operators (to allow making histogram data point movable)
